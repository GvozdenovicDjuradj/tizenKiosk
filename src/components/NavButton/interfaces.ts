import { ViewStyle } from "react-native"
import { ButtonProps } from "../Button"

export interface Styles {
  navButton: ViewStyle
  navButtonLeft: ViewStyle
  navButtonRight: ViewStyle
  triangleLeft: ViewStyle
  triangleRight: ViewStyle
  view: ViewStyle
}

export interface NavButtonProps extends ButtonProps {
  buttonColor?: string
  privacyPolicyButtonDefaultText?: string
}

export interface NavButtonNextState {
  isDisabled: boolean
}
