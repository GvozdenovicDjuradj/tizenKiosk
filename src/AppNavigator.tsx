import React from "react"
import { BackHandler } from "react-native"
import { connect } from "react-redux"
import { createStackNavigator, NavigationInjectedProps } from "react-navigation"
import { AppScreens, RootState } from "./interfaces"
import { selectors } from "./utils"
import ConfirmationScreen from "./components/confirmationScreen"
import CustomerScreen from "./components/customerScreen"
import Home from "./components/Home"
import ServiceSelectionScreen from "./components/serviceSelection"
import SubserviceSelectionScreen from "./components/serviceSelection/SubServiceSelection"
import ServiceInfoScreen from "./components/serviceSelection/ServiceInfoScreen"
import NoConnectionModal from "./components/NoConnectionModal"
import CheckInScreen from "./components/checkInScreen"
import CheckInConfirmation from "./components/checkInConfirmationScreen"
import EventCheckInScreen from "./components/eventCheckInScreen"
import EventCheckInConfirmationScreen from "./components/eventCheckInConfirmationScreen"
import KioskClosedScreen from "./components/kioskClosedScreen"
import QuestionsScreen from "./components/QuestionsScreen"
import QueueFullOrNAScreen from "./components/QueueFullOrNAScreen"
import ScreenSaver from "./components/screenSaver"
import { handleBackButton, removeBackButtonHandler } from "./utils/backButton"
import PrivacyPolicyScreen from "./components/privacyPolicyScreen"
import KeepAwake from 'react-native-keep-awake'
import QueueUnderCapacityScreen from "./components/QueueUnderCapacityScreen"

const AppStackNavigator = createStackNavigator({
  [AppScreens.HOME]: { screen: Home },
  [AppScreens.SERVICE_SELECTION]: { screen: ServiceSelectionScreen },
  [AppScreens.SUBSERVICE_SELECTION]: { screen: SubserviceSelectionScreen },
  [AppScreens.SERVICE_INFO]: { screen: ServiceInfoScreen },
  [AppScreens.CUSTOMER_DETAILS]: { screen: CustomerScreen },
  [AppScreens.SCREEN_SAVER]: { screen: ScreenSaver },
  [AppScreens.QUEUE_CONFIRMATION]: {
    screen: ConfirmationScreen,
    navigationOptions: {
      gesturesEnabled: false,
    },
  },
  [AppScreens.QUEUE_FULL_OR_NA]: { screen: QueueFullOrNAScreen },
  [AppScreens.QUEUE_UNDER_OCCUPANCY]: { screen: QueueUnderCapacityScreen },
  [AppScreens.QUESTIONS]: { screen: QuestionsScreen },
  [AppScreens.CHECK_IN]: { screen: CheckInScreen },
  [AppScreens.EVENT_CHECK_IN]: { screen: EventCheckInScreen },
  [AppScreens.EVENT_CHECK_IN_CONFIRMATION]: { screen: EventCheckInConfirmationScreen },
  [AppScreens.NO_CONNECTION_MODAL]: { screen: NoConnectionModal },
  [AppScreens.KIOSK_CLOSED]: { screen: KioskClosedScreen },
  [AppScreens.CHECK_IN_CONFIRMATION]: {
    screen: CheckInConfirmation,
    navigationOptions: {
      gesturesEnabled: false,
    },
  },
  [AppScreens.PRIVACY_POLICY]: { screen: PrivacyPolicyScreen },
}, {
  cardStyle: { backgroundColor: "transparent" },
  headerMode: "none",
  initialRouteName: AppScreens.HOME,
  mode: "card",
})

interface Props extends NavigationInjectedProps {
  loading: boolean,
  hasKioskModeEnable: boolean
}

class AppNavigator extends React.Component<Props> {
  public static router = AppStackNavigator.router

  /**
   * Depending on hasKioskModeEnable props, calling back button will Quit the app or do nothing
   * react navigation back button handler will overwrite that behavior
   */
  public handleBackButtonClick = () => {
   if (!this.props.hasKioskModeEnable) {
     BackHandler.exitApp()
   }
  }

  public componentWillMount() {
    handleBackButton(this.handleBackButtonClick)
  }

  public componentWillUnmount() {
    removeBackButtonHandler(this.handleBackButtonClick)
  }

  public render() {
    const { hasKioskModeEnable } = this.props
    return (
      <React.Fragment>
        { hasKioskModeEnable &&
          <KeepAwake />
        }
        <AppStackNavigator navigation={this.props.navigation} />
      </React.Fragment>
    )
  }
}

const mapStateToProps = (state: RootState) => ({
  hasKioskModeEnable: selectors.isKioskModeEnabled(state)
})

export default connect(mapStateToProps)(AppNavigator)
