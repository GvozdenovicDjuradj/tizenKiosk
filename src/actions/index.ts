export * from "./app"
export * from "./error"
export * from "./kiosk"
export * from "./kioskSettings"
export * from "./printing"
export * from "./navigationActions"
export * from "./screenSaver"
