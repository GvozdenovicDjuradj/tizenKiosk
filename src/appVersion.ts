import { Platform } from "react-native"
import DeviceInfo from "react-native-device-info"

export default Platform.select({
  web: "n/a",
  default: DeviceInfo.getReadableVersion()
})
