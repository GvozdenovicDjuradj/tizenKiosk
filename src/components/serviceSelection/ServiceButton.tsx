import React from "react"
import {
  Platform,
  StyleProp,
  Text,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from "react-native"
import moment from "moment"
import "moment-duration-format"
import { connect } from "react-redux"
import { AnyProduct } from "../../interfaces/product"
import { mixins } from "../../theme"
import { RootState, KioskSettings } from "../../interfaces"
import { metrics } from "../../utils"
import { NavigationEvents } from "react-navigation"
import { getTranslatedProductName } from '../../utils/selectors'

export interface StateProps { 
  isDisabled: boolean
}

interface PropsFromState {
  serviceButtonColor: string;
  textColor: string;
  fontFamily: string;
  productName?: string;
  isLoading: boolean;
}

interface OwnProps {
  onProductPress: (product: AnyProduct) => any
  product: AnyProduct
  showWaitTime?: boolean;
  style?: any,
  textStyle?: any
  withIcon?: boolean
}

interface ServiceButtonProps extends PropsFromState, OwnProps { }

const buttonStyle: StyleProp<ViewStyle> = {
  alignItems: "center",
  elevation: 7,
  justifyContent: "center",
  margin: 15,
  padding: 15,
  shadowOffset: { width: 2, height: 5 },
  shadowOpacity: 0.8,
  shadowRadius: 2,
  width: "30%",
  ...mixins.borderRadius(6),
}

const iconText = {
  fontFamily: Platform.OS === "ios" ? "FontAwesome5FreeSolid" : "fontawesome-solid",
}

const viewStyle: ViewStyle = {
  alignItems: "center",
  flexDirection: "row",
  justifyContent: "center",
}

const textStyle: TextStyle = {
  fontSize: 24,
  textAlign: "center",
}
/**
 * Returns string in format `{days} d {hours} h {minutes} min`
 * @param time time to parse in minutes
 */
const formatTime = (time = 0): string => moment
  .duration(time, "minutes")
  .format("d [d] h [h] m [min]", { trim: "both" })

class ProductButton extends React.Component<ServiceButtonProps, StateProps> {
  public state = {
    isDisabled: false
  }

  UNSAFE_componentWillReceiveProps(nextProps: ServiceButtonProps) {
    // In case there's an error `isLoading` is going to be set to false and we want to enable the button again
    if (!nextProps.isLoading) {
      this.setState({isDisabled: false});
    }
  }

  public render() {
    const {
      onProductPress,
      product,
      serviceButtonColor,
      showWaitTime = false,
      style,
      textColor,
      fontFamily,
      withIcon = false,
      productName,
      isLoading,
    } = this.props
    const buttonStyles = [
      buttonStyle,
      style,
      { height: metrics.minSize(150, metrics.vh(18))},
      { backgroundColor: serviceButtonColor, shadowColor: "#000" },
    ];

    return (
      <React.Fragment>
        <NavigationEvents
          onWillFocus={this.disableHandler(false)}
          onWillBlur={this.disableHandler(true)}
        />
        <TouchableOpacity
          activeOpacity={0.75}
          onPress={() => { 
              this.setState({isDisabled: true});  
              onProductPress(product); 
          }}
          style={buttonStyles}
          disabled={this.state.isDisabled || isLoading}
        >
          <View style={viewStyle}>
            {withIcon ? (
              <View style={{ paddingHorizontal: 10 }}>
                <Text style={[iconText, { color: textColor }]}>&#xf019;</Text>
              </View>
            ) : null}
            <Text numberOfLines={2} style={[textStyle, { color: textColor, fontFamily }]}>
              {productName}
            </Text>
          </View>
          {showWaitTime ? (
            <View style={viewStyle}>
              <Text style={[textStyle, { color: textColor, fontSize: 18, fontFamily }]}>
                Wait time: {formatTime(product.waitTime)}
              </Text>
            </View>
          ) : null}
        </TouchableOpacity>
      </React.Fragment>
    )
  }

  private readonly disableHandler = (isDisabled: boolean) => () => {
    this.setState({
      isDisabled
    })
  }
}

const mapStateToProps = (state: RootState, props: OwnProps) => {
  const settings = state.kiosk.settings as KioskSettings
  return {
    serviceButtonColor: settings && settings.template.serviceButtonColor || "#000",
    textColor: settings && settings.template.buttonTextColor || "#fff",
    fontFamily: settings && settings.template.font || "Arial",
    productName: getTranslatedProductName(state, props.product),
    isLoading: state.kiosk.isAddingCustomerToQueue,
  }
}

export default connect(mapStateToProps)(ProductButton)
