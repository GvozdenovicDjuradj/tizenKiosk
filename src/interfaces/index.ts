export * from "./answers"
export * from "./app"
export * from "./appState"
export * from "./checkIn"
export * from "./customerInQueue"
export * from "./eventCheckIn"
export * from "./kiosk"
export * from "./modal"
export * from "./navigation"
export * from "./product"
export * from "./question"
export * from "./routes"
export * from "./utils"
export * from "./validation"
export * from "./venue"
export * from "./screenSaver"

import { AppState } from "./app"
import { CheckInState } from "./checkIn"
import { EventCheckInState } from "./eventCheckIn"
import { KioskSettingsForm } from "../reducers/kioskSettings"
import { KioskState } from "./kiosk"
import { ModalState } from "./modal"
import { QuestionAnswer } from "./answers"
import { QuestionsReducer } from "./question"
import { ValidationState } from "./validation"
import { PrintingState } from "./printing"
import { PrinterState } from "../printer/reducer"
import { ScreenSaverState } from "./screenSaver"

export interface Flags {
  [key: string]: any
}

export type ValidatorFn = (text: string, ...args: any[]) => { error: string }

export interface RootState {
  answers: QuestionAnswer[];
  app: AppState;
  checkIn: CheckInState;
  error: string;
  eventCheckIn: EventCheckInState;
  kiosk: KioskState;
  kioskSettings: KioskSettingsForm;
  modal: ModalState;
  printer: PrinterState;
  printing: PrintingState;
  questions: QuestionsReducer;
  validation: ValidationState;
  screenSaver?: ScreenSaverState;
}
