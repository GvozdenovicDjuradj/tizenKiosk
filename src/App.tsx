import React, { Component } from "react"
import { Store } from "redux"
import { Provider } from "react-redux"
import { StatusBar, View, PermissionsAndroid, Platform } from "react-native"
import { PersistGate } from "redux-persist/integration/react"
import { Navigator as NavigatorType, setNavigator } from "./navigation"
import Navigator from "./routes"
import { applicationStart, initApp as initAppActionCreator, stopApp } from "./actions/app"
import { RootState } from "./interfaces"
import { persistor, store } from "./store"
import { setup } from "./utils/localeString"
import Modals from "./modals"
import Printing from "./components/Printing"
import Ticket from "./printer/containers/Ticket"
import Notification from "./components/Notification"
import initSentry from "./initSentry"
import { getKioskIdAndDomainFromUrl } from "./actions"
import FadeNotification from "./printer/containers/FadeNotification"

if (Platform.OS !== "web") {
    initSentry()
}

setup(store)

if (Platform.OS === "web") {
    setTimeout(() => {
        store.dispatch(getKioskIdAndDomainFromUrl())
    }, 0)
}

setTimeout(() => store.dispatch(applicationStart()))

const start = (store$: Store<RootState>, navigator: NavigatorType) => {
    if (navigator) {
        setNavigator(navigator)
        store$.dispatch(initAppActionCreator())

    } else {
        store$.dispatch(stopApp())
    }
}

export default class App extends Component {
    public async componentDidMount() {
        if (Platform.OS === "android") {
            // TODO - This should be re-prompted in settings later, I will handle this in a later ticket
            await PermissionsAndroid.request(
              PermissionsAndroid.PERMISSIONS.READ_EXTERNAL_STORAGE,
              {
                title: "Qudini Kiosk - Printing",
                message: `We require access to read a configuration
                file which stores internal settings for the connected
                printer, in addition we also need to read saved tickets
                which we save to external storage as an image.`,
                buttonNegative: "Decline",
                buttonPositive: "Accept",
              },
            )

            await PermissionsAndroid.request(
              PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE,
              {
                title: "Qudini Kiosk - Printing",
                message: `We require access to write to a configuration
                  file which stores internal settings for the connected
                  printer, in addition we also need to write saved tickets
                  which we save to external storage as an image.`,
                buttonNegative: "Decline",
                buttonPositive: "Accept",
              },
            )
        }
    }

    public render() {
        return (
            <Provider store={store}>
                <PersistGate persistor={persistor}>
                    <View style={{ flex: 1 }}>
                        <StatusBar hidden/>
                        <Ticket/>
                        <View style={{ flex: 1, width: "100%", height: "100%", overflow: "hidden" }}>
                            <Modals/>
                            <Notification/>
                            <Navigator ref={(navigator: NavigatorType) => start(store, navigator)}/>
                            <Printing />
                        </View>
                        <FadeNotification />
                    </View>
                </PersistGate>
            </Provider>
        )
    }
}
