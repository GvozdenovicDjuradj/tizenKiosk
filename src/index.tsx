import "babel-polyfill"
import { AppRegistry } from "react-native"
import App from "./App"
import app from "../app.json"

AppRegistry.registerComponent(app.name, () => App)

if (window && typeof window === "object" && window.document) {
  AppRegistry.runApplication(app.name, { rootTag: document.getElementById("app") })
}
