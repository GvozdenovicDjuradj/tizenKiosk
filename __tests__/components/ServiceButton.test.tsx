import React from "react"
import renderer from "react-test-renderer"
import { mount } from "enzyme"
import { Provider } from "react-redux"
import { TouchableOpacity } from "react-native"
import mockStore from "redux-mock-store"
import ServiceButton from "../../src/components/serviceSelection/ServiceButton"
import { AnyProduct } from "../../src/interfaces"
import { createOnPressEvent } from "../fixtures"

describe("Service button tests", () => {
  const store = mockStore()({
    kiosk: {
      settings: {
        template: {
          buttonTextColor: "#ffffff",
          serviceButtonColor: "#ff0000",
        }
      }
    }
  })

  it("Should render ServiceButton component successfully", () => {
    const product: AnyProduct = {
      name: "product"
    }
    const pressFn = () => ({ })
    const tree = renderer.create(
      <Provider store={store}>
        <ServiceButton onProductPress={pressFn} product={product} />
      </Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("Should pass props to ServiceButton successfully", () => {
    const product: AnyProduct = {
      waitTime: 4455,
      title: "product tile",
    }
    const pressFn = jest.fn()
    const tree = renderer.create(
      <Provider store={store}>
        <ServiceButton
          onProductPress={pressFn}
          product={product}
          showWaitTime
          withIcon
        />
      </Provider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it("Should call function passed", () => {
    const product: AnyProduct = {
      name: "product"
    }
    const pressFn = jest.fn()
    const wrapper = mount(
      <Provider store={store}>
        <ServiceButton onProductPress={pressFn} product={product} />
      </Provider>
    )
    const props = wrapper.find(TouchableOpacity).at(0).props()
    if (props.onPress) {
      props.onPress(createOnPressEvent())
    }
    expect(pressFn).toBeCalledWith(product)
  })

})
