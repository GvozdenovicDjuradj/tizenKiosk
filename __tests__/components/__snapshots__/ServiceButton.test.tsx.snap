// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service button tests Should pass props to ServiceButton successfully 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ff0000",
      "borderBottomLeftRadius": 6,
      "borderBottomRightRadius": 6,
      "borderTopLeftRadius": 6,
      "borderTopRightRadius": 6,
      "elevation": 7,
      "height": 150,
      "justifyContent": "center",
      "margin": 15,
      "opacity": 1,
      "padding": 15,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 5,
        "width": 2,
      },
      "shadowOpacity": 0.8,
      "shadowRadius": 2,
      "width": "30%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "FontAwesome5FreeSolid",
            },
            Object {
              "color": "#ffffff",
            },
          ]
        }
      >
        ï€™
      </Text>
    </View>
    <Text
      numberOfLines={2}
      style={
        Array [
          Object {
            "fontSize": 24,
            "textAlign": "center",
          },
          Object {
            "color": "#ffffff",
            "fontFamily": "Arial",
          },
        ]
      }
    >
      product tile
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 24,
            "textAlign": "center",
          },
          Object {
            "color": "#ffffff",
            "fontFamily": "Arial",
            "fontSize": 18,
          },
        ]
      }
    >
      Wait time: 
      3 d 2 h 15 mins
    </Text>
  </View>
</View>
`;

exports[`Service button tests Should render ServiceButton component successfully 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ff0000",
      "borderBottomLeftRadius": 6,
      "borderBottomRightRadius": 6,
      "borderTopLeftRadius": 6,
      "borderTopRightRadius": 6,
      "elevation": 7,
      "height": 150,
      "justifyContent": "center",
      "margin": 15,
      "opacity": 1,
      "padding": 15,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 5,
        "width": 2,
      },
      "shadowOpacity": 0.8,
      "shadowRadius": 2,
      "width": "30%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      numberOfLines={2}
      style={
        Array [
          Object {
            "fontSize": 24,
            "textAlign": "center",
          },
          Object {
            "color": "#ffffff",
            "fontFamily": "Arial",
          },
        ]
      }
    >
      product
    </Text>
  </View>
</View>
`;
